yamux/src/connection/stream/flow_control.rs

```diff
@@ -0,0 +1,305 @@
+    }
+
+    fn assert_invariants(&self, buffer_len: usize) {
+        if !cfg!(debug_assertions) {
+            return;
+        }
+
+        let config = &self.config;
+        let rtt = self.rtt.get();
+        let accumulated_max_stream_windows = *self.accumulated_max_stream_windows.lock();
+
+        assert!(
+            buffer_len <= self.max_receive_window_size as usize,
+            "The current buffer size never exceeds the maximum stream receive window."
+        );
+        assert!(
+            self.current_receive_window_size <= self.max_receive_window_size,
+            "The current window never exceeds the maximum."
+        );
+        assert!(
+            (self.max_receive_window_size - DEFAULT_CREDIT) as usize
+                <= config.max_connection_receive_window.unwrap_or(usize::MAX)
+                    - config.max_num_streams * DEFAULT_CREDIT as usize,
+            "The maximum never exceeds its maximum portion of the configured connection limit."
+        );
+        assert!(
+            (self.max_receive_window_size - DEFAULT_CREDIT) as usize
+                <= accumulated_max_stream_windows,
+            "The amount by which the stream maximum exceeds DEFAULT_CREDIT is tracked in accumulated_max_stream_windows."
+        );
+        if rtt.is_none() {
+            assert_eq!(
+                self.max_receive_window_size, DEFAULT_CREDIT,
+                "The maximum is only increased iff an rtt measurement is available."
+            );
+        }
+    }
+
+    pub(crate) fn current_send_window_size(&self) -> u32 {
+        self.current_send_window_size
+    }
+
+    pub(crate) fn consume_send_window(&mut self, i: u32) {
+        self.current_send_window_size = self
+            .current_send_window_size
+            .checked_sub(i)
+            .expect("not exceed send window");
+    }
+
+    pub(crate) fn increase_send_window_by(&mut self, i: u32) {
+        self.current_send_window_size = self
+            .current_send_window_size
+            .checked_add(i)
+            .expect("send window not to exceed u32");
+    }
+
+    pub(crate) fn current_receive_window_size(&self) -> u32 {
+        self.current_receive_window_size
+    }
+
+    pub(crate) fn consume_receive_window(&mut self, i: u32) {
+        self.current_receive_window_size = self
+            .current_receive_window_size
+            .checked_sub(i)
+            .expect("not exceed receive window");
+    }
+}
+
+impl Drop for FlowController {
+    fn drop(&mut self) {
+        let mut accumulated_max_stream_windows = self.accumulated_max_stream_windows.lock();
+
+        debug_assert!(
+            *accumulated_max_stream_windows
+                >= (self.max_receive_window_size - DEFAULT_CREDIT) as usize,
+            "{accumulated_max_stream_windows} {}",
+            self.max_receive_window_size
+        );
+
+        *accumulated_max_stream_windows -= (self.max_receive_window_size - DEFAULT_CREDIT) as usize;
+    }
+}
```

Comment:
This also doesn't look like it should be part of the algorithm but live in `Stream` instead.

Reply:
I would like to bundle all flow-control related logic in one module. Thus I prefer keeping this logic in `flow_control` instead of `stream`.